File: php/plugins/doppler.php
<?php

/**
 * Doppler, as a voxgig/plugin definition.
 *
 * A PLUGIN, not a built-in: it opens a socket.
 */

declare(strict_types=1);

namespace Voxgig\Sekreto\Plugins;

require_once __DIR__ . '/../src/Providers.php';
require_once __DIR__ . '/../src/Addr.php';
require_once __DIR__ . '/httpjson.php';

use Voxgig\Sekreto\Name;
use Voxgig\Sekreto\Provider;
use Voxgig\Sekreto\SekretoError;

use function Voxgig\Sekreto\checkaddr;
use function Voxgig\Sekreto\providerplugin;

/**
 * Doppler.
 *
 * The whole config is downloaded once - Doppler's own bulk endpoint - and
 * answered from memory, like a remote .env: `api.token` is the
 * `API_TOKEN` entry. A service token is config-scoped, so project and
 * config are only needed with broader tokens.
 */
class DopplerProvider implements Provider
{
    /** @var array<string, string>|null */
    private ?array $values = null;

    /** @param array<string, mixed> $opts */
    public function __construct(private array $opts = [])
    {
    }

    /** @return array<string, string> */
    private function load(): array
    {
        if (null !== $this->values) {
            return $this->values;
        }

        $addr = rtrim(($this->opts['addr'] ?? null) ?: 'https://api.doppler.com', '/');
        checkaddr($addr);

        $url = $addr . '/v3/configs/config/secrets/download?format=json';
        if ($this->opts['project'] ?? null) {
            $url .= '&project=' . rawurlencode($this->opts['project']);
        }
        if ($this->opts['config'] ?? null) {
            $url .= '&config=' . rawurlencode($this->opts['config']);
        }

        [$status, $body] = fetchjson(
            'GET',
            $url,
            ['Authorization: Bearer ' . ($this->opts['token'] ?? '')]
        );

        if (200 !== $status || !is_array($body)) {
            throw new SekretoError('sekreto: doppler error: ' . $status);
        }

        $this->values = [];
        foreach ($body as $key => $value) {
            if (null !== $value) {
                $this->values[$key] = (string) $value;
            }
        }

        return $this->values;
    }

    public function lookup(string $name): ?string
    {
        return $this->load()[Name::envkey($name)] ?? null;
    }

    public function describe(): string
    {
        $project = $this->opts['project'] ?? null;

        return 'doppler'
            . ($project ? ':' . $project . '/' . ($this->opts['config'] ?? '') : '');
    }
}

/**
 * The `doppler` provider kind, as a voxgig/plugin definition. Pass it to
 * `Sekreto` in the `plugins` option; nothing else loads it.
 *
 * @return array<string, mixed>
 */
function doppler(): array
{
    return providerplugin('doppler', fn(array $spec) => new DopplerProvider($spec));
}

File: csharp/plugins/Doppler.cs
// Doppler, as a voxgig/plugin definition.
//
// PLUGIN CODE. This file is in the VoxgigSekretoPlugins assembly, which
// the core does not reference - so nothing here is linked into an
// application whose chain names only built-in kinds. See
// docs/design/plugin-providers.md.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Globalization;
using System.IO;
using System.Net.Http;
using System.Text;

using Definition = Voxgig.Plugin.Definition;

namespace Voxgig.Sekreto.Plugins
{
    /// <summary>
    /// Doppler.
    ///
    /// <para>The whole config is downloaded once - Doppler's own bulk
    /// endpoint - and answered from memory, like a remote .env: `api.token`
    /// is the `API_TOKEN` entry. A service token is config-scoped, so
    /// project and config are only needed with broader tokens.</para>
    /// </summary>
    public class DopplerProvider : IProvider
    {
        private readonly string token;
        private readonly string project;
        private readonly string config;
        private readonly string addr;

        private Dictionary<string, string> values;

        public DopplerProvider(string token = null, string project = null,
            string config = null, string addr = null)
        {
            this.token = token;
            this.project = project;
            this.config = config;
            this.addr = addr;
        }

        private Dictionary<string, string> Load()
        {
            if (null != values)
            {
                return values;
            }

            string useaddr = (string.IsNullOrEmpty(addr) ? "https://api.doppler.com" : addr)
                .TrimEnd('/');
            Addr.Check(useaddr);

            string url = useaddr + "/v3/configs/config/secrets/download?format=json";
            if (!string.IsNullOrEmpty(project))
            {
                url += "&project=" + Url.Encode(project);
            }
            if (!string.IsNullOrEmpty(config))
            {
                url += "&config=" + Url.Encode(config);
            }

            var (status, body) = Http.FetchJson("GET", url,
                new Dictionary<string, string> { ["authorization"] = "Bearer " + (token ?? "") });

            if (200 != status || !(body is Dictionary<string, object> entries))
            {
                throw new SekretoError("sekreto: doppler error: " + status);
            }

            values = new Dictionary<string, string>();
            foreach (var entry in entries)
            {
                if (null != entry.Value)
                {
                    values[entry.Key] = Convert.ToString(entry.Value);
                }
            }

            return values;
        }

        public string Lookup(string name)
        {
            return Load().TryGetValue(Names.EnvKey(name), out string found) ? found : null;
        }

        public string Describe()
        {
            return "doppler"
                + (string.IsNullOrEmpty(project) ? "" : ":" + project + "/" + (config ?? ""));
        }
    }

    /// <summary>The <c>doppler</c> provider kind.</summary>
    public static class Doppler
    {
        public static readonly Definition Plugin = Providers.ProviderPlugin(
            "doppler", spec => new DopplerProvider(
                Providers.Text(spec.GetValueOrDefault("token")),
                Providers.Text(spec.GetValueOrDefault("project")),
                Providers.Text(spec.GetValueOrDefault("config")),
                Providers.Text(spec.GetValueOrDefault("addr"))));
    }
}
